### PR TITLE
Enable auto-validation in Web Lab

### PR DESCRIPTION
**Pull request by jmkulwik**
This is part of a larger feature to enable auto-validation for all CSD and CSP levels where a code change is expected.

There are two things that needed to happen in order to enable auto-validation in Web Lab

1. Web Lab needed to support error dialogs (in the past Web Lab would progress to the next level no matter what and would display no success/failure dialog)
2. Web Lab needed a diff tool that would allow us to determine whether the code had changed from the initial code on the level.

As part of this change, I also enabled auto-validation on one level in Web Lab corresponding to https://studio.code.org/s/csd2-2020/stage/2/puzzle/2

### Error dialog
The feedback module is what controls error and success dialog. In the past, feedback.js assumed that an app was either droplet or blockly and would follow code paths accordingly. Since bramble is fundamentally different than both, the would cause errors to be thrown. To handle this, I added another constant to the pageConstants. isBlockly and isDroplet already existed. I added isBramble so conditions could be evaluated on this case.

Additionally, StudioApp, from where the feedback dialog is created, expects all apps to use the init function and create a StudioApp config. Web Lab and other non-droplet/blockly apps do not do this, however. (Currently, Fish/Oceans is the only other app that does not use init or the StudioApp config). Since the feedback feature was previously never used by Web Lab, I needed to adjust it to no longer require the StudioApp config.

### Diff Tool
Other apps use project.js to manage their code and asset storage. For those other apps, project.js knows about their code and can perform a diff against the level code. Web Lab, however, uses bramble. To diff the web lab code, we need to use bramble. In this PR, additional functionality is added to bramble to get all files from the file system and serialize them to JSON (the original form that the files are in when the level is created). From here, we can diff the existing code from the original level code.

*Note:* Images are passed to Bramble as URLs but are stored as data strings. It is not possible to go from a data string back to the original URL. Therefore, rather than diffing the data in an image, we instead check for a different image name. This diff tool, therefore, cannot be used to reliably diff images - only code. We will only enable it on levels where code changes are expected or where a student is expected to change image names.

## Links

- [spec](https://docs.google.com/document/d/1hF-wvav0MXLnH49x0WBeO8TcUibMaCUegZiDr8-6l-c/edit#)
- [Dev-Doc](https://docs.google.com/document/d/1xUgybM56dV5kmqOukKVPW1d88a9HGhphUs2r0tWlC_4/edit#)
- [LP-1551](https://codedotorg.atlassian.net/browse/LP-1551)
- Unblocks [LP-1549](https://codedotorg.atlassian.net/browse/LP-1549)
- Starts [LP-1550](https://codedotorg.atlassian.net/browse/LP-1550)

## Testing story

I went ahead and added tests for WebLab.js. There are, however, significant problems with brambleHost that require a refactor before testing can happen. Internal folks can see this slack thread for a discussion around these problems: https://codedotorg.slack.com/archives/C1B3PNDL7/p1603143998153300

Copying some of the details here:
* brambleHost relies on multiple window variables that are added by something else in the dependency chain. Example: window.requirejs.config. As a first step, I set up some if(IN_UNIT_TEST) conditions to handle these, but I ran into a second problem.
* Functions are not exported. brambleHost, instead, injects its exports directly into WebLab with webLab_.setBrambleHost. webLab_ is created by a call to parent.getWebLab() - I couldn't find an easy way to mock this, so I set a couple more if(IN_UNIT_TEST) conditions as a temporary option, but I ran into yet another problem.
* Functions can't be mocked. It was very convenient while testing WebLab.js to mock some properties in order to test just one specific function. brambleHost example: validateProjectChanged calls getAllFileDataFromBramble. I would have liked to mock getAllFileDataFromBramble, but since these aren't set up as class properties, it wasn't possible. Granted, getAllFileDataFromBramble is really more of a private function, so maybe I shouldn't be trying to mock it anyway, but one can wish :slightly_smiling_face: That being said, I ran into a final problem and finally decided to throw in the towel.
* bramble_, which gets used in many places, is injected by a load operation. This then ends up with a bunch of side effects which would require a lot of setup for a simple unit test. At this point, I realized the best path forward would be to refactor brambleHost (the first idea that came to mind was to move it to a class model and make the global variables/functions class properties).

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/weblab/brambleHost.js [hunk 1]
```diff
@@ -302,6 +302,66 @@ function syncFilesWithBramble(fileEntries, currentProjectVersion, callback) {
...
+    }
+
+    const changedFile = startSources.files.find(startFile => {
+      const matchingFile = userFiles.files.find(
+        file => file.name === startFile.name
+      );
+      // If startFile doesn't have `data` (and instead has something different
+      // like `url`), this is an image and is stored differently in the
+      // startSources than in bramble. Check for a matching file name, but
+      // don't compare data.
```
Comment (breville) [inline apps/src/weblab/brambleHost.js:322]: Does this refer to startFile containing the string `data` vs. `url`, or is it more like binary content vs. alphanumeric content?

Comment (jmkulwik) [reply apps/src/weblab/brambleHost.js:322]: Both. If startFile contains 'url' then that means the source in bramble will be stored as 'data' with binary content. If startFile contains 'data' then that means the source in bramble will be stored as 'data' with alphanumeric content.

File: apps/src/weblab/brambleHost.js [hunk 2]
```diff
@@ -302,6 +302,66 @@ function syncFilesWithBramble(fileEntries, currentProjectVersion, callback) {
...
+      return (
+        !matchingFile ||
+        (startFile.data &&
+          startFile.data.replace(/\s+/g, '') !==
+            matchingFile.data.replace(/\s+/g, ''))
```
Comment (breville) [inline apps/src/weblab/brambleHost.js:328]: It might be worth a comment on what is happening here, maybe with an example?

**Comment by jmkulwik** [comment]
UI tests passed here but unit tests ran into a transient seeding issue: https://drone.cdn-code.org/code-dot-org/code-dot-org/562

File: apps/src/weblab/brambleHost.js (hunk 2 above)
Comment (jmkulwik) [reply apps/src/weblab/brambleHost.js:328]: Added a comment! It just removes whitespace.

**Review by breville — APPROVED** [review]
This looks great.  Hopefully somebody with more knowledge of Web Lab can also review, but this looks well thought-through and the PR write-up is really effective.

**Review by cforkish — APPROVED** [review]
LGTM!

**Pull request merged**